src/OSPSuite.Core/Domain/Services/ConcurrencyManager.cs

```diff
@@ -38,56 +38,99 @@ public interface IConcurrencyManager
       Task<IDictionary<TData, ConcurrencyManagerResult<TResult>>> RunAsync<TData, TResult>(
          IReadOnlyList<TData> data,
-         Func<TData, CancellationToken, TResult> action,
+         Func<TData, CancellationToken, TResult> func,
          CancellationToken cancellationToken,
          int numberOfCoresToUse
       ) where TData : IWithId;
+
+      /// <summary>
+      /// </summary>
+      /// <typeparam name="TData">Data type to consume by the worker function</typeparam>
+      /// <param name="data">List of data to consume by the workers</param>
+      /// <param name="action">
+      ///    An action running on each worker on each piece of data.
+      /// </param>
+      /// <param name="cancellationToken">Cancellation token to cancel the threads</param>
+      /// <param name="numberOfCoresToUse">Number of cores to use. Use null to take all cores</param>
+      /// <returns>Dictionary binding a result for each input data after running the action on it</returns>
+      Task RunAsync<TData>(
+         IReadOnlyList<TData> data,
+         Action<TData, CancellationToken> action,
+         CancellationToken cancellationToken,
+         int numberOfCoresToUse
+      );
    }
 
    public class ConcurrencyManager : IConcurrencyManager
    {
+      private readonly IObjectTypeResolver _objectTypeResolver;
+
+      public ConcurrencyManager(IObjectTypeResolver objectTypeResolver)
+      {
+         _objectTypeResolver = objectTypeResolver;
+      }
+
       public async Task<IDictionary<TData, ConcurrencyManagerResult<TResult>>> RunAsync<TData, TResult>
       (
          IReadOnlyList<TData> data,
-         Func<TData, CancellationToken, TResult> action,
+         Func<TData, CancellationToken, TResult> func,
          CancellationToken cancellationToken,
          int numberOfCoresToUse
       ) where TData : IWithId
       {
          var results = new ConcurrentDictionary<TData, ConcurrencyManagerResult<TResult>>();
 
+         verifyUniqueIdsAreUsed(data);
+
          await Task.Run(() => Parallel.ForEach(data, createParallelOptions(cancellationToken, numberOfCoresToUse),
-               datum =>
+            datum =>
+            {
+               cancellationToken.ThrowIfCancellationRequested();
+               try
+               {
+                  results.TryAdd(
+                     datum,
+                     new ConcurrencyManagerResult<TResult>(datum.Id, func(datum, cancellationToken))
+                  );
+               }
+               catch (Exception e)
                {
-                  cancellationToken.ThrowIfCancellationRequested();
-                  try
-                  {
-                     results.TryAdd(
-                        datum, 
-                        new ConcurrencyManagerResult<TResult>(datum.Id, action(datum, cancellationToken))
-                     );
-                  }
-                  catch (Exception e)
```

Comment:
You removed the try/catch. This was here to make sure that if one run fails (running one simulation throws an exception for example) the rest are still recorded. Like this I think you would not be able to return anything at all. Right?

Reply:
yep. But here we have no way of knowing which one failed as we have no identifier.

Reply:
Fair enough.